storage/cluster-unification: multi-process storage clusters

guswynn opened the pull request:
Closes https://github.com/MaterializeInc/materialize/issues/17419

This pr closes the above issue, _and has a carefully implementation_, such that the changes do not block the type of evolution we want for storage clusters as discussed in https://github.com/MaterializeInc/materialize/issues/17419. The implementation is effectively:

- Add a `CreateTimely` variant to `StorageCommand`. The new `mz-cluster-client` crate holds some code here that used by both `StorageCommad` and `ComputeCommand`
- Merge the `mz_storage::server` and `mz_compute::server` modules, in the new `mz-cluster` crate. This primarily handles `CreateTimely` commands, before passing commands off to either `storage_state` or `compute_state`.
- Change storage rehydration to use `connect_partitioned` instead of `connect`.

As noted in https://materializeinc.slack.com/archives/CM7ATT65S/p1675105911234109, the new crate structure is roughly:
```
            mz-cluster-client
            /        \
mz-storage-client   mz-compute-client
   |             \    |
   |  mz-cluster  \   |
   |     |       \ |  |
   mz_storage     mz_compute
          \       /
          clusterd
```

This pr _is primarily code movement_, but:
- Some the code movement, mostly around `mz_cluster::server` is quite complex, and has some subtle generics
- The protocols for compute and storage still differ (primarily in how internal command flow works), so some of the partitioned implementations differ in subtle ways
  - I have added comments in places where these differences need to be carefully considered

I highly recommend the `Tips for Reviewer` section before reviewing. Also, CI will likely NOT be kind, I will be working on it continuously.


### Motivation


  * This PR adds a known-desirable feature.

### Tips for reviewer
I have attempted to use logical commits to make this pr easier to review, I recommend reviewing them separately:

- The first commit just adds a basic `mz-cluster-client` crate
- The second commit splits the `TimelyConfig` and `ClusterStartupEpoch` into `mz-cluster-client`
- The third commit adds a basic `mz-cluster` crate
- The fourth commit moves the `mz_compute::server` module into `mz-cluster` and makes the implementation more generic
- The fifth commit is the primary implementation of the feature. It adapts `mz_storage::server` into the `mz_cluster::server` module and implements `CreateTimely` support on the client (controller) side. There is quite a broad commit
- The 6th commit switches the `unsafe` STORAGECTL ADDRESS` feature into `STORAGECTL ADDRESSES`+`STORAGE ADDRESSES`, and fixes up all the tests that use this feature
- The 7th commit removes the unused `--storage-workers` flag from clusterd. @philip-stoev I *think* I found all the places that needed `WORKERS 4` added to them to get back to the old behavior. The only other semantic changes this causes is:
  - for unmanaged, compute clusters, the (unused) timely cluster will have 1 thread instead of 4 on each replica process
  - The `KafkaScenario` and `PgCdcScenario` in `bounded-memory` will be running on 1 thread instead of 4. Let me know if you want this fixed. That would require making a `clusterd` and `storaged` cluster, instead of just 1, or making the compute queries run on 4 threads instead of 1.
  - Anything else that runs `CREATE SOURCE/SINK .. IN CLUSTER` on a default-sized cluster will be running on a `4-4` cluster, but this might not even happen anywhere
  - Otherwise, `--default-storage-host-size` should keep everything the same
- The 8th commit fixes resizing storage replicas
- The 9th commit removes some unused deps
- The 10th commit ensure we handle replica epochs correctly
- The 11th commit adjusts some existing tests to ensure we have multi-process source test coverage

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - [ ] I still need to add explicit tests for this new feature, the branch currently just attempts to get things working
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

umanwizard commented on src/cluster/src/communication.rs:
```diff
@@ -0,0 +1,341 @@
+//!
+//! The startup protocol is as follows:
+//! The compute controller in `environmentd`, after having connected to all the
+//! `clusterd` processes in a replica, sends each of them a `CreateTimely` command
```
It would be helpful if you made the delete of this file from compute part of the same commit, so it can be verified that it's not changing.

guswynn replied:
Ill see what I can do; im not a git expert

guswynn replied:
wowww just removing it in the commit below, git figures outs its a move+edit, thats amazing

umanwizard commented on src/cluster-client/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "mz-cluster-client"
-description = "The public API for both compute and storage"
+description = "Types for storage and compute."
```
Why's this changing? `cluster-client` isn't just for types, is it?

guswynn replied:
ill clean this up

umanwizard commented on src/cluster-client/src/lib.rs:
```diff
@@ -74,6 +74,6 @@
 // END LINT CONFIG
 
-//! Code that storage AND compute need.
+//! Types for storage and compute.
```
same here

umanwizard commented on src/cluster/src/types.rs:
```diff
@@ -0,0 +1,50 @@
+            cmd => Err(cmd),
+        }
+    }
+}
```
It's not immediately clear what these traits do; they could use docs.

teskje replied:
Could we also `#[warn(missing_docs)]` on the new crates? It's quite annoying to add that after the fact.

(Since the new crates contain existing code "it's not my job do document other people's code" would be a valid response!)

umanwizard commented on src/compute/src/server.rs:
```diff
@@ -11,307 +11,81 @@
 
-/// Configures a dataflow server.
-#[derive(Debug)]
+/// Configures the
```
Unfinished comment here

umanwizard commented on src/cluster/src/server.rs:
```diff
@@ -0,0 +1,302 @@
+
+/// Configures a dataflow server.
+#[derive(Debug)]
+pub struct GenericConfig {
```
Optional nit -- "Generic" to me usually implies Rust generics, which this struct isn't. Maybe another name like `SharedConfig` would be better? (And then a comment to say what it's shared between). Or `BaseConfig` ?

benesch replied:
Oh, _this_ is what you meant by generic! I was confused too. Recommend `ClusterConfig`.

umanwizard reviewed (commented):
I only looked at the first few commits (the ones that are mostly related to compute). The general idea seems sane to me.

benesch commented on src/cluster/src/server.rs:
```diff
@@ -0,0 +1,303 @@
+
+type PartitionedClient<C, R, A> = Partitioned<LocalClient<C, R, A>, C, R>;
+
+/// Configures a dataflow server.
```
```suggestion
/// Configures a cluster server.
```

benesch commented on src/cluster/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+[package]
```
Does this need to be a separate crate from `clusterd`? I'd probably just move the `clusterd` binary into this crate.

guswynn replied:
@benesch this crate is _above_ the storage and compute crates, so it can't be merged

benesch reviewed (commented):
The fact that this passes CI is 🔥.

I didn't review in depth, and leaving the full review to the experts, but the basic principles look 👍🏽 to me. The one thing that wasn't clear to me was: do we have sufficient test coverage of horizontally scaled storage clusters? This will be on by default in prod for the customers who have access to 2xlarge+ replicas, so it should be well tested before we ship this.

philip-stoev commented:
Nightly does not look very good against this branch, e.g.
 ```
zippy-storaged-1         | thread 'tokio-runtime-worker' panicked at 'Reference to absent shard 1 for collection u8', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-00028e6de25859d35-1/materialize/tests/src/storage-client/src/client.rs:606:33
 ```

Please let me know if you need any help tracking the reasons for individual failures. I will hold off testing this PR until Nightly gets greener. Thanks!

philip-stoev commented:
And another one:

```
zippy-materialized-1     | thread 'coordinator' panicked at 'sinks should be validated by sequence_create_sink: Catalog(Error { kind: Stash(StashError { inner: Other("unexpected peek multiplicity") }) })', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-00028e6de25859d35-1/materialize/tests/src/adapter/src/coord/message_handler.rs:486:22

```
in https://buildkite.com/materialize/nightlies/builds/1753#0186114c-3a5a-46a8-9c6b-645061d531cf

petrosagg commented on src/cluster/src/types.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+pub trait CreatableCommand {
+    fn is_create_timely(self) -> Result<(TimelyConfig, ClusterStartupEpoch), Self>
```
Not sure if this was considered but I would find a parent enum much easier to reason about:

```rust
enum ClusterCommand {
    CreateTimely(...),
    StorageCommand(StorageCommand),
    ComputeCommand(ComputeCommand),
}
```
Would something like this work? It will also allow us to move more common commands up in the future.

If we end up keeping this trait, I think we need a better name because `is_foo` methods are usually returning bool. `try_into_create_timely` would be one option, or somehow integrating with the `TryFrom` traits.

guswynn replied:
A parent enum is a significant amount of code changes, that I am specifically avoiding until we merge the command streams!

the `try_into_` is great! ill probably keep the special trait but rename things

petrosagg commented on src/storage-client/src/client.rs:
```diff
@@ -483,7 +503,7 @@ where
 
     fn new(parts: usize) -> PartitionedStorageState<T> {
         PartitionedStorageState {
-            parts: parts.try_into().expect("more than 4 billion partitions"),
```
Changing this to `usize` requires doing `as` conversions below and add lints, do we need to change it?

guswynn replied:
because `compute` uses `usize`, so I made them the same; I felt more comfortable changing storage code, ill add a todo to make them both u32's

guswynn replied:
added a todo!

petrosagg commented on src/storage-client/src/client.rs:
```diff
@@ -494,16 +514,27 @@ where
-                            Some(Antichain::from_elem(T::minimum()));
-                            usize::cast_from(self.parts)
-                        ];
+                        // TODO(benesch): fix this dangerous use of `as`.
```
If you want to keep this it should probably be `TODO(gus)` :D

guswynn replied:
its true!

petrosagg commented on src/storage/src/storage_state.rs:
```diff
@@ -863,6 +880,9 @@ impl<'w, A: Allocate> Worker<'w, A> {
         for command in &mut commands {
             match command {
+                StorageCommand::CreateTimely { .. } => {
+                    panic!("CreateTimely must be captured before")
```
This panic would go away with the common `ClusterCommand` parent enum

petrosagg commented on src/controller/src/clusters.rs:
```diff
@@ -425,6 +436,12 @@ where
                             port_hint: 2100,
                         },
+                        // To simplify the changes to tests, the port
+                        // chosen here is _after_ the compute ones.
```
Can you add a `TODO(petrosagg)` to fix it? It bothers be enough that I'm willing to do it after this gets merged

guswynn replied:
thanks for offering!

petrosagg commented on src/storage-client/src/client.rs:
```diff
@@ -627,7 +627,13 @@ where
+                            // this state, because it reconciles it state as soon as we connect,
+                            // so we can just ignore `DroppedIds` for collections we are _currently
+                            // in the process of recreating_.
+                            continue;
```
In the past not having strict checks here had hidden subtle bugs. The comment seems to imply there is something stricter we can do here instead of unconditionally continuing. Can we check if the `id` in question is in the set of "currently being created" and still panic if not?

guswynn replied:
Hmm, I need to think about this!

guswynn replied:
This was broken by an erroneous change added in a previous commit! Its fixed!

petrosagg reviewed (commented):
Did a first pass and overall it looks good! I think before I do a second round of review it would be good to discuss this in a call with you just to make sure I understand how everything works.

Also, can you add a short explanation (either as a reply here or in the code) of why we need the `Activatable` thing and what needs to be done to remove it?

teskje commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -239,63 +245,48 @@ where
+                config: timely_config,
+                // The replica epoch is unused by storage, and won't
+                // be used till multi-replica storage is supported.
+                epoch: ClusterStartupEpoch::new(envd_epoch, 0),
```
Don't we need to count up the replica epoch even with just a single replica? It is used to ensure that the connections between processes get initialized correctly and we don't run into dead-locks when a strict subset of the replica processes crashes during initialization. @umanwizard has the most context on what can go wrong.

umanwizard replied:
Yes, good catch. This number needs to increase every time we attempt to launch the same replica.

guswynn replied:
This epoch is _never_ checked in the storage code, its only checked in the `mz_compute_client::controller::{instance, replica}` modules, but i can bump it in this pr for clarity sake

teskje replied:
The point is that the storage code *should* probably check the epoch if we want to support multi-process storage clusters, to avoid running into the same kind of split-brain issues that Brennan had to deal with before he added that epoch.

The comment makes it sound like the epoch is needed to support multi-replica clusters, but that's not the case. Multiple replicas can get by fine without that.

umanwizard replied:
> This epoch is _never_ checked in the storage code

Then what are you filling in the `CreateTimely` command that gets sent to clusterd (and ultimately ends up in `communication.rs` ?

umanwizard replied:
It doesn't need to be checked in any of the envd-side code; it needs to be properly maintained on that side, and then used in communication.rs

umanwizard replied:
As for what can go wrong otherwise, nobody really knows in detail, there are just a million difficult-to-reason-about edge cases. That’s why I ripped out the old logic and replaced it with this epoch checking.

But the main theme is that in a multi-process cluster, if some processes fail to connect and others succeed, the failed ones might be rebooted and respond to a _future_ CreateTimely command. Then you get into a situation where some of the processes have responded to one command and others to another. That’s why now we made them check that they all have the same epoch in CreateTimely and otherwise refuse to let them build a timely instance together. So, every CreateTimely that is instantiated for a given replica needs to have a strictly increasing epoch.

teskje reviewed (commented):
The changes look mostly good to me. I have a bunch of nits and one more serious concern about not counting up the replica epoch.

jkosh44 commented on src/cluster-client/Cargo.toml:
```diff
@@ -0,0 +1,54 @@
+[package]
+name = "mz-cluster-client"
```
Does this crate need to be added to `CODEOWNERS`?

jkosh44 commented on src/sql-parser/tests/testdata/ddl:
```diff
@@ -1135,11 +1135,11 @@ CREATE CLUSTER REPLICA default.replica IDLE ARRANGEMENT MERGE EFFORT = 0
+CREATE CLUSTER REPLICA default.replica STORAGECTL ADDRESSES = ('1', '2'), COMPUTECTL ADDRESSES = ('1', '2'), COMPUTE ADDRESSES = ('3', '4'), WORKERS = 2, INTROSPECTION INTERVAL = NULL
 =>
-CreateClusterReplica(CreateClusterReplicaStatement { of_cluster: Ident("default"), definition: ReplicaDefinition { name: Ident("replica"), options: [ReplicaOption { name: StoragectlAddress, value: Some(Sequence([Value(String("1")), Value(String("2"))])) }, ReplicaOption { name: ComputectlAddresses, value: Some(Sequence([Value(String("1")), Value(String("2"))])) }, ReplicaOption { name: ComputeAdd ...
+CreateClusterReplica(CreateClusterReplicaStatement { of_cluster: Ident("default"), definition: ReplicaDefinition { name: Ident("replica"), options: [ReplicaOption { name: StoragectlAddresses, value: Some(Sequence([Value(String("1")), Value(String("2"))])) }, ReplicaOption { name: ComputectlAddresses, value: Some(Sequence([Value(String("1")), Value(String("2"))])) }, ReplicaOption { name: ComputeA ...
```
Can we add a test with the `STORAGE ADDRESSES` option?

jkosh44 commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -1192,8 +1192,10 @@ pub enum ReplicaOptionName {
     AvailabilityZone,
-    /// The `STORAGECTL ADDRESS` option.
-    StoragectlAddress,
+    /// The `STORAGE ADDRESSES` option.
+    StorageAddresses,
+    /// The `STORAGECTL ADDRESSES` option.
+    StoragectlAddresses,
```
Are these behind unsafe mode? If not we probably want to update the docs with the new syntax.

guswynn replied:
Yes they are behind unsafe mode!

jkosh44 reviewed (approved):
Had a couple of minor nits/questions for the surfaces code, otherwise LGTM. I haven't looked at any of the non-surfaces code.

guswynn commented:
> The one thing that wasn't clear to me was: do we have sufficient test coverage of horizontally scaled storage clusters? This will be on by default in prod for the customers who have access to 2xlarge+ replicas, so it should be well tested before we ship this.

No, I need to add more tests for this; Likely its a basic testdrive test as well as a `test/cluster` test that manually sets up 2 clusterds and 

> Also, can you add a short explanation (either as a reply here or in the code) of why we need the Activatable thing and what needs to be done to remove it?

Compute uses `SyncActivator`s to unpark the timely threads, so that their custom dataflow operator (on worker 0) can pull off new commands to distribute. Storage uses `Thread::unpark` and pulls the commands off on all workers, then only worker 0 happens to distribute them to the sequencer.

The compute team has indicated that they prefer the way storage uses `Sequencer`, so we can remove the `A: Awaitable` stuff once we move them to the storage technique. This consolidation will be required anyways, when we merge the command streams.

guswynn commented on src/cluster/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+tokio = { version = "1.24.2", features = ["fs", "rt", "sync", "net"] }
+tracing = "0.1.37"
+uuid = { version = "1.2.2", features = ["serde", "v4"] }
+workspace-hack = { version = "0.0.0", path = "../workspace-hack" }
```
will clean these up!

guswynn replied:
Gonna do this as a 9th commit to make this the easiest to manage

guswynn commented on src/cluster/src/server.rs:
```diff
@@ -0,0 +1,303 @@
+/// Threadsafe reference to an optional TimelyContainer
+pub type TimelyContainerRef<C, R, A> = Arc<tokio::sync::Mutex<Option<TimelyContainer<C, R, A>>>>;
+
+/// Initiates a timely dataflow computation, processing compute commands.
```
Probably `ClusterCommand` and `ClusterResponse`

guswynn commented:
Fixed everyone's nits and commits, except for the replica epoch bumping; will add that tomorrow, running a new nightly over tonight

philip-stoev commented:
Item No 1.  The test/limits test fails with

```
pg8000.dbapi.ProgrammingError: {'S': 'ERROR', 'C': '42601', 'M': 'Expected right square bracket, found ADDRESSES', 'P': '288'}
```

so I guess there is some more syntax to be adjusted?

philip-stoev commented:
Item No 2 . Can you please rebase again so that you get https://github.com/MaterializeInc/materialize/pull/17502 and then schedule a new Nightly run. Thank you!

philip-stoev commented:
Item No 3. The CI is still using "4" for the storage clusters, even though it should start using "4-4" or at least "4-1". I made a change in my own branch to 4-4 and am now running the Buildkite pipelines.

philip-stoev commented:
Item No 4. The "Platform Checks Upgrade" tests are failing because the syntax is not compatible between the old and the new version. At this time there is no workaround for that situation, so those tests will keep failing until the next release is out :-(

guswynn commented:
> Item No 1. The test/limits test fails with

I couldn't for the life of me find this error, thanks!

> Item No 3. The CI is still using "4" for the storage clusters, even though it should start using "4-4" or at least "4-1". I made a change in my own branch to 4-4 and am now running the Buildkite pipelines.

Can we do thats as a followup?

> Item No 4. The "Platform Checks Upgrade" tests are failing because the syntax is not compatible between the old and the new version. At this time there is no workaround for that situation, so those tests will keep failing until the next release is out :-(

Yeah its real unfortunate; we will need to think about how to avoid this kind of problem in the future

philip-stoev reviewed (approved):
I got a 4-4 run going in Nightly, and there were a bunch of unrelated panics, OOMs and timeouts but no obvious specific regressions:

https://buildkite.com/materialize/nightlies/builds/1762#0186174a-1f8b-4325-8c09-71e946dab9c9

So I am hitting Accept here. Gus agreed that we will aim to to have 4-4 or at least 4-1 by default in CI before the next refactoring in the area. I will open a follow up PR to change the default in the CI and will open any product issues as tickets.

guswynn commented:
Updated and should be _entirely_ ready for a final review:

- @philip-stoev I added a commit that tests that we actually distribute work across workers, and adjust `test/cluster` to test multi-process sources by default
- @umanwizard @teskje I added a commit ensure we bump the replica epoch
  - Note that I bump it _every time_ we reconnect to the underlying cluster! This means by default, when sources/sinks with linked clusters come up, the replica might be bumped a few times as the cluster is starting up! I am not too concerned by this, especially as we phase out linked-clusters.

guswynn commented:
Running one more nightly: https://buildkite.com/materialize/nightlies/builds/1774 to ensure the epoch-handling stuff doesn't regress things!

umanwizard commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -281,11 +281,17 @@ where
             let timely_command = StorageCommand::CreateTimely {
                 config: timely_config,
                 // The replica epoch is unused by storage, and won't
                 // be used till multi-replica storage is supported.
```
I think we can remove this comment

umanwizard reviewed (commented):
The epoch tracking looks fine to me!

teskje commented on .github/CODEOWNERS:
```diff
@@ -27,6 +27,8 @@
 /src/build-info                     @benesch
 /src/ccsr                           @benesch
 /src/cloud-resources                @MaterializeInc/cloud
+/src/cluster                       @MaterializeInc/compute @MaterializeInc/storage
```
Nit: The indentation looks is off here.

teskje reviewed (approved):
LGTM!

aljoscha commented on src/cluster-client/src/client.rs:
```diff
@@ -0,0 +1,216 @@
+    pub ctl_addrs: Vec<String>,
+    /// The network addresses of the dataflow (Timely) endpoints for
+    /// each process in the replica.
+    pub dataflow_addrs: Vec<String>,
```
I know this is not your PR, but I had to look up what exactly these mean. Might be good to document that as a follow-up. My understanding is:
 - ctl_addrs is for the controller talking to the "cluster"
 - dataflow_addrs is for internal networking, such that the different processes that make up a cluster can find and talk to each other
Is that correct?

guswynn replied:
I will look at this as a followup!

aljoscha commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -150,21 +154,21 @@ struct RehydrationTask<T> {
     AwaitAddress,
     /// The storage replica should be (re)hydrated.
     Rehydrate {
         /// The network address of the storage replica.
```
nit: rustdoc could be updated here, and below. But it's not super important for now.

aljoscha commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -226,29 +234,70 @@ where
+                //
+                // TODO(guswynn): cluster-unification: ensure this is cleaned up when
+                // the compute and storage command streams are merged.
+                idle_arrangement_merge_effort: 1337,
```
:sweat_smile:

aljoscha commented on src/storage/src/server.rs:
```diff
@@ -9,38 +9,38 @@
-    pub persist_clients: Arc<PersistClientCache>,
+
+    /// Metrics for sources.
+    pub source_metrics: SourceBaseMetrics,
```
It's nice how these `Configs` for compute and storage are cleaned up!

guswynn replied:
yes!

aljoscha commented on src/storage/src/storage_state.rs:
```diff
@@ -379,8 +387,17 @@ impl<'w, A: Allocate> Worker<'w, A> {
             match self.client_rx.recv() {
-                Ok((rx, tx)) => self.run_client(rx, tx),
-                Err(_) => shutdown = true,
+                // We don't use the activator sender here, because we use a `Sequencer` to
```
I don't think this comment is correct: we _do_ send back the current thread (which is the timely thread of this worker), because that's what's reading from the command_tx (which then pushes commands into the internal flow).

The code is correct, it seems, though.

aljoscha commented on test/cluster/mzcompose.py:
```diff
@@ -431,7 +439,8 @@ def workflow_test_github_15930(c: Composition) -> None:
                 logging_on (
-                    STORAGECTL ADDRESS 'clusterd1:2100',
+                    STORAGECTL ADDRESSES ['clusterd1:2100'],
+                    STORAGE ADDRESSES ['clusterd1:2103'],
```
must have been nice, fixing up all these... :see_no_evil:

guswynn replied:
yes!

aljoscha commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -215,10 +215,16 @@ where
-
         // Reconnect to the storage replica.
-        let client = {
+        let stream = Retry::default()
```
nit: maybe `retry_stream`?

guswynn replied:
I am going to leave this one as it was!

aljoscha commented on src/storage-client/src/controller/rehydration.rs:
```diff
@@ -281,11 +281,15 @@ where
             };
 
+            // The first epoch we actually send to the cluster will be `1`, just like compute.
+            let new_epoch = ClusterStartupEpoch::new(
```
We do also verify this epoch business for storage clusters because that code is shared/generic now, right?

https://github.com/MaterializeInc/materialize/blob/cd31ae1cb71af9adccb0bac67f8aac851eacc1d5/src/cluster/src/server.rs#L291

aljoscha commented on test/testdrive/basic-multi-process-storage.td:
```diff
@@ -0,0 +1,71 @@
+# by the Apache License, Version 2.0.
+
+# Tests that multi-process storage objects (in this case, sources)
+# actually utilize both all the processes in the replica.
```
both or all? :sweat_smile:

aljoscha commented on test/testdrive/basic-multi-process-storage.td:
```diff
@@ -0,0 +1,71 @@
+# actually utilize both all the processes in the replica.
+
+# Create a topic with enough partitions and data spread
+# among them so have a very, very high likelihood
```
typo: to have

aljoscha commented on test/testdrive/basic-multi-process-storage.td:
```diff
@@ -0,0 +1,71 @@
+# NOTE: These queries are slow to succeed because the default metrics scraping
+# interval is 30 seconds.
+
+# Until compute and storage and _entirely unified_, storage will not have
```
typo: are _entirely unified_

aljoscha reviewed (approved):
It's excellent! I like how neatly this pulls the cluster business out of compute and storage, and therefore makes the compute/storage-specific things tighter. I had some comments/nits/questions.

Also, this was :chefs-kiss: to review because of splitting into logical commits.

guswynn commented on src/cluster/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+[package]
+name = "mz-cluster"
+description = "Tools for storage and compute."
+version = "0.38.0-dev"
```
Yes! I will update this!
